File: wrappers/java/src/main/java/org/hyperledger/indy/sdk/cache/Cache.java
package org.hyperledger.indy.sdk.cache;

import com.sun.jna.Callback;
import org.hyperledger.indy.sdk.IndyException;
import org.hyperledger.indy.sdk.IndyJava;
import org.hyperledger.indy.sdk.LibIndy;
import org.hyperledger.indy.sdk.ParamGuard;
import org.hyperledger.indy.sdk.pool.Pool;
import org.hyperledger.indy.sdk.wallet.Wallet;

import java.util.concurrent.CompletableFuture;

/**
 * cache.rs API
 */

/**
 * High level wrapper around did SDK functions.
 */
public class Cache extends IndyJava.API {
    private Cache() {

    }

    /*
	 * STATIC CALLBACKS
	 */

	/**
	 * Callback used when a function returning Void completes.
	 */
	private static Callback voidCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err) {

			CompletableFuture<Void> future = (CompletableFuture<Void>) removeFuture(xcommand_handle);
			if (! checkResult(future, err)) return;

			Void result = null;
			future.complete(result);
		}
	};

	/**
	 * Callback used when a function returning String completes.
	 */
	private static Callback stringCb = new Callback() {

		@SuppressWarnings({"unused", "unchecked"})
		public void callback(int xcommand_handle, int err, String str) {

			CompletableFuture<String> future = (CompletableFuture<String>) removeFuture(xcommand_handle);
			if (! checkResult(future, err)) return;

			String result = str;
			future.complete(result);
		}
	};

	/*
	 * STATIC METHODS
	 */

	/**
	 * Purge schema cache.
	 *
	 * EXPERIMENTAL
	 *
	 * @param wallet      The wallet.
	 * @param optionsJson The record tags used for search and storing meta information as json:
	 *                    {
	 *                        "maxAge": -1, // (int, optional, -1 by default) Purge cached data if older than this many seconds. -1 means purge all.
	 *                    }
	 * @return A future that resolves no value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<Void> purgeSchemaCache(
			Wallet wallet,
			String optionsJson) throws IndyException {

		ParamGuard.notNull(wallet, "wallet");
		ParamGuard.notNull(optionsJson, "optionsJson");

		CompletableFuture<Void> future = new CompletableFuture<Void>();
		int commandHandle = addFuture(future);

		int walletHandle = wallet.getWalletHandle();

		int result = LibIndy.api.indy_purge_schema_cache(
				commandHandle,
				walletHandle,
				optionsJson,
				voidCb);

		checkResult(future, result);

		return future;
	}

	/**
	 * Purge credential definition cache.
	 *
	 * EXPERIMENTAL
	 *
	 * @param wallet      The wallet.
	 * @param optionsJson The record tags used for search and storing meta information as json:
	 *                    {
	 *                        "maxAge": -1, // (int, optional, -1 by default) Purge cached data if older than this many seconds. -1 means purge all.
	 *                    }
	 * @return A future that resolves no value.
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<Void> purgeCredDefCache(
			Wallet wallet,
			String optionsJson) throws IndyException {

		ParamGuard.notNull(wallet, "wallet");
		ParamGuard.notNull(optionsJson, "optionsJson");

		CompletableFuture<Void> future = new CompletableFuture<Void>();
		int commandHandle = addFuture(future);

		int walletHandle = wallet.getWalletHandle();

		int result = LibIndy.api.indy_purge_cred_def_cache(
				commandHandle,
				walletHandle,
				optionsJson,
				voidCb);

		checkResult(future, result);

		return future;
	}

    /**
	 * Gets schema json data for specified schema id.
	 * If data is present inside of cache, cached data is returned.
	 * Otherwise data is fetched from the ledger and stored inside of cache for future use.
	 *
	 * EXPERIMENTAL
	 *
	 * @param pool           The pool.
	 * @param wallet         The wallet.
	 * @param submitterDid   DID of the submitter stored in secured Wallet
	 * @param id             The id of schema.
	 * @param optionsJson
	 *  {
	 *    noCache: (optional, false by default) Skip usage of cache,
	 *    noUpdate: (optional, false by default) Use only cached data, do not try to update.
	 *    noStore: (optional, false by default) Skip storing fresh data if updated
	 *    minFresh: (optional, -1 by default) Return cached data if not older than this many seconds. -1 means do not check age.
	 *  }
	 * @return A future that resolves to Schema json:
	 * {
	 *     id: identifier of schema
	 *     attrNames: array of attribute name strings
	 *     name: Schema's name string
	 *     version: Schema's version string
	 *     ver: Version of the Schema json
	 * }
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<String> getSchema(
	        Pool pool,
			Wallet wallet,
			String submitterDid,
			String id,
			String optionsJson) throws IndyException {

		ParamGuard.notNull(pool, "pool");
		ParamGuard.notNull(wallet, "wallet");
		ParamGuard.notNull(submitterDid, "submitterDid");
		ParamGuard.notNull(id, "id");
		ParamGuard.notNull(optionsJson, "optionsJson");

		CompletableFuture<String> future = new CompletableFuture<String>();
		int commandHandle = addFuture(future);

        int poolHandle = pool.getPoolHandle();
		int walletHandle = wallet.getWalletHandle();

		int result = LibIndy.api.indy_get_schema(
				commandHandle,
				poolHandle,
				walletHandle,
				submitterDid,
				id,
				optionsJson,
				stringCb);

		checkResult(future, result);

		return future;
	}

    /**
	 * Gets credential definition json data for specified credential definition id.
	 * If data is present inside of cache, cached data is returned.
	 * Otherwise data is fetched from the ledger and stored inside of cache for future use.
	 *
	 * EXPERIMENTAL
	 *
	 * @param pool           The pool.
	 * @param wallet         The wallet.
	 * @param submitterDid   DID of the submitter stored in secured Wallet
	 * @param id             The id of credential definition.
	 * @param optionsJson
	 *  {
	 *    noCache: (optional, false by default) Skip usage of cache,
	 *    noUpdate: (optional, false by default) Use only cached data, do not try to update.
	 *    noStore: (optional, false by default) Skip storing fresh data if updated
	 *    minFresh: (optional, -1 by default) Return cached data if not older than this many seconds. -1 means do not check age.
	 *  }
	 * @return A future that resolves to Credential Definition json:
	 * {
	 *     id: string - identifier of credential definition
	 *     schemaId: string - identifier of stored in ledger schema
	 *     type: string - type of the credential definition. CL is the only supported type now.
	 *     tag: string - allows to distinct between credential definitions for the same issuer and schema
	 *     value: Dictionary with Credential Definition's data: {
	 *         primary: primary credential public key,
	 *         Optional[revocation]: revocation credential public key
	 *     },
	 *     ver: Version of the Credential Definition json
	 * }
	 * @throws IndyException Thrown if an error occurs when calling the underlying SDK.
	 */
	public static CompletableFuture<String> getCredDef(
	        Pool pool,
			Wallet wallet,
			String submitterDid,
			String id,
			String optionsJson) throws IndyException {

		ParamGuard.notNull(pool, "pool");
		ParamGuard.notNull(wallet, "wallet");
		ParamGuard.notNull(submitterDid, "submitterDid");
		ParamGuard.notNull(id, "id");
		ParamGuard.notNull(optionsJson, "optionsJson");

		CompletableFuture<String> future = new CompletableFuture<String>();
		int commandHandle = addFuture(future);

        int poolHandle = pool.getPoolHandle();
		int walletHandle = wallet.getWalletHandle();

		int result = LibIndy.api.indy_get_cred_def(
				commandHandle,
				poolHandle,
				walletHandle,
				submitterDid,
				id,
				optionsJson,
				stringCb);

		checkResult(future, result);

		return future;
	}
}

File: wrappers/rust/src/cache.rs
use futures::Future;

use {ErrorCode, IndyError};

use std::ffi::CString;

use utils::callbacks::{ClosureHandler, ResultHandler};

use ffi::cache;
use ffi::{ResponseEmptyCB, ResponseStringCB};
use {WalletHandle, CommandHandle, PoolHandle};

/// Get schema json data for specified schema id.
/// If data is present inside of cache, cached data is returned.
/// Otherwise data is fetched from the ledger and stored inside of cache for future use.
///
/// EXPERIMENTAL
///
/// # Arguments
/// * `pool_handle` - pool handle (created by open_pool_ledger).
/// * `wallet_handle` - wallet handle (created by open_wallet).
/// * `submitter_did` - DID of the submitter stored in secured Wallet.
/// * `id` - identifier of schema.
/// * `options_json` -
///  {
///    noCache: (bool, optional, false by default) Skip usage of cache,
///    noUpdate: (bool, optional, false by default) Use only cached data, do not try to update.
///    noStore: (bool, optional, false by default) Skip storing fresh data if updated,
///    minFresh: (int, optional, -1 by default) Return cached data if not older than this many seconds. -1 means do not check age.
///  }
/// # Returns
/// Schema json.
/// {
///     id: identifier of schema
///     attrNames: array of attribute name strings
///     name: Schema's name string
///     version: Schema's version string
///     ver: Version of the Schema json
/// }
pub fn get_schema(pool_handle: PoolHandle,
                  wallet_handle: WalletHandle,
                  submitter_did: &str,
                  id: &str,
                  options_json: &str) -> Box<dyn Future<Item=String, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_string();

    let err = _get_schema(command_handle, pool_handle, wallet_handle, submitter_did, id, options_json, cb);

    ResultHandler::str(command_handle, err, receiver)
}

pub fn _get_schema(command_handle: CommandHandle,
                   pool_handle: PoolHandle,
                   wallet_handle: WalletHandle,
                   submitter_did: &str,
                   id: &str,
                   options_json: &str,
                   cb: Option<ResponseStringCB>) -> ErrorCode {
    let submitter_did = c_str!(submitter_did);
    let id = c_str!(id);
    let options_json = c_str!(options_json);

    ErrorCode::from(
        unsafe {
            cache::indy_get_schema(command_handle, pool_handle, wallet_handle, submitter_did.as_ptr(), id.as_ptr(), options_json.as_ptr(), cb)
        }
    )
}

/// Get credential definition json data for specified credential definition id.
/// If data is present inside of cache, cached data is returned.
/// Otherwise data is fetched from the ledger and stored inside of cache for future use.
///
/// EXPERIMENTAL
///
/// # Arguments
/// * `pool_handle` - pool handle (created by open_pool_ledger).
/// * `wallet_handle` - wallet handle (created by open_wallet).
/// * `submitter_did` - DID of the submitter stored in secured Wallet.
/// * `id` - identifier of credential definition.
/// * `options_json` -
///  {
///    noCache: (bool, optional, false by default) Skip usage of cache,
///    noUpdate: (bool, optional, false by default) Use only cached data, do not try to update.
///    noStore: (bool, optional, false by default) Skip storing fresh data if updated,
///    minFresh: (int, optional, -1 by default) Return cached data if not older than this many seconds. -1 means do not check age.
///  }
/// # Returns
/// Credential Definition json.
/// {
///     id: string - identifier of credential definition
///     schemaId: string - identifier of stored in ledger schema
///     type: string - type of the credential definition. CL is the only supported type now.
///     tag: string - allows to distinct between credential definitions for the same issuer and schema
///     value: Dictionary with Credential Definition's data: {
///         primary: primary credential public key,
///         Optional<revocation>: revocation credential public key
///     },
///     ver: Version of the Credential Definition json
/// }
pub fn get_cred_def(pool_handle: PoolHandle,
                    wallet_handle: WalletHandle,
                    submitter_did: &str,
                    id: &str,
                    options_json: &str) -> Box<dyn Future<Item=String, Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec_string();

    let err = _get_cred_def(command_handle, pool_handle, wallet_handle, submitter_did, id, options_json, cb);

    ResultHandler::str(command_handle, err, receiver)
}

pub fn _get_cred_def(command_handle: CommandHandle,
                     pool_handle: PoolHandle,
                     wallet_handle: WalletHandle,
                     submitter_did: &str,
                     id: &str,
                     options_json: &str,
                     cb: Option<ResponseStringCB>) -> ErrorCode {
    let submitter_did = c_str!(submitter_did);
    let id = c_str!(id);
    let options_json = c_str!(options_json);

    ErrorCode::from(
        unsafe {
            cache::indy_get_cred_def(command_handle, pool_handle, wallet_handle, submitter_did.as_ptr(), id.as_ptr(), options_json.as_ptr(), cb)
        }
    )
}

/// Purge schema cache.
///
/// EXPERIMENTAL
///
/// # Arguments
/// * `wallet_handle` - wallet handle (created by open_wallet).
/// * `id` - identifier of schema.
/// * `options_json` -
///  {
///    maxAge: (int, optional, -1 by default) Purge cached data if older than this many seconds. -1 means purge all.
///  }
pub fn purge_schema_cache(wallet_handle: WalletHandle, options_json: &str) -> Box<dyn Future<Item=(), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec();

    let err = _purge_schema_cache(command_handle, wallet_handle, options_json, cb);

    ResultHandler::empty(command_handle, err, receiver)
}

fn _purge_schema_cache(command_handle: CommandHandle, wallet_handle: WalletHandle, options_json: &str, cb: Option<ResponseEmptyCB>) -> ErrorCode {
    let options_json = c_str!(options_json);

    ErrorCode::from(unsafe { cache::indy_purge_schema_cache(command_handle, wallet_handle, options_json.as_ptr(), cb) })
}

/// Purge credential definition cache.
///
/// EXPERIMENTAL
///
/// # Arguments
/// * `wallet_handle` - wallet handle (created by open_wallet).
/// * `id` - identifier of credential definition.
/// * `options_json` -
///  {
///    maxAge: (int, optional, -1 by default) Purge cached data if older than this many seconds. -1 means purge all.
///  }
pub fn purge_cred_def_cache(wallet_handle: WalletHandle, options_json: &str) -> Box<dyn Future<Item=(), Error=IndyError>> {
    let (receiver, command_handle, cb) = ClosureHandler::cb_ec();

    let err = _purge_cred_def_cache(command_handle, wallet_handle, options_json, cb);

    ResultHandler::empty(command_handle, err, receiver)
}

fn _purge_cred_def_cache(command_handle: CommandHandle, wallet_handle: WalletHandle, options_json: &str, cb: Option<ResponseEmptyCB>) -> ErrorCode {
    let options_json = c_str!(options_json);

    ErrorCode::from(unsafe { cache::indy_purge_cred_def_cache(command_handle, wallet_handle, options_json.as_ptr(), cb) })
}